File: search-incomplete-results/python/github_search_incomplete.py
"""Say whether a GitHub search is being answered in part and nobody noticed.

Read only. One GET per round against /search/*, three rounds by default, with a
pause between them. Nothing is written.

Search runs against a server-side timeout. When a query outruns it GitHub does
not fail: it returns what it found with the top-level incomplete_results boolean
set to true. Status 200, valid JSON, fewer items, no error anywhere. A client
that reads items and ignores the flag treats a partial answer as a complete one,
and the count quietly moves between runs.

This is not the 1,000-result ceiling. That limit is deterministic, announces
itself with a 422 when you page past it, and applies to queries with more than a
thousand matches. This flag is non-deterministic, arrives on the first page, and
fires on queries nowhere near the cap. total_count is read here purely so the
ceiling can be ruled out by name.

Search has its own rate bucket, separate from and much tighter than core, so
this check is deliberately tiny and refuses a plan that would not fit in it.

Environment:

    GITHUB_TOKEN    any token with read access; search needs authentication to
                    get the larger of the two search buckets
"""
import argparse
import json
import logging
import os
import re
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("github_search_incomplete")

API = "https://api.github.com"
UA = "github-search-incomplete/1.0"

# The other Search limit, read only so it can be excluded as an explanation.
RESULT_CAP = 1000
# Authenticated search requests per minute. The check is sized against this.
SEARCH_BUCKET = 30

QUALIFIER = re.compile(r"(?:^|\s)-?([A-Za-z_]+):\S")

SCOPES = ("repo", "org", "user")
RANGES = ("created", "updated", "merged", "closed")


def flagged(body):
    """Whether this response says it is partial. Pure."""
    return bool(isinstance(body, dict) and body.get("incomplete_results") is True)


def total_of(body):
    """The reported match count, or None. Pure."""
    if not isinstance(body, dict):
        return None
    try:
        return int(body.get("total_count"))
    except (TypeError, ValueError):
        return None


def item_count(body):
    """How many items were actually delivered. Pure."""
    if not isinstance(body, dict):
        return 0
    items = body.get("items")
    return len(items) if isinstance(items, list) else 0


def cacheable(body):
    """Whether this response may be stored. Pure.

    The rule the whole note comes down to: a partial answer written to a cache
    without a note of its partiality becomes the permanent truth.
    """
    return isinstance(body, dict) and not flagged(body)


def above_result_cap(total):
    """Whether the ceiling could also be in play here. Pure."""
    try:
        return int(total) > RESULT_CAP
    except (TypeError, ValueError):
        return False


def qualifiers(query):
    """The qualifier names used in a search query. Pure."""
    return set(QUALIFIER.findall(" " + str(query or "")))


def narrowing(query):
    """Which narrowing devices the query is not already using. Pure."""
    have = qualifiers(query)
    out = []
    if not have & set(SCOPES):
        out.append("repo: or org:")
    if not have & set(RANGES):
        out.append("created: or updated: date range")
    if "language" not in have:
        out.append("language:")
    return out


def observe(body):
    """The three fields worth keeping from one response. Pure."""
    return {"incomplete": flagged(body),
            "total": total_of(body),
            "items": item_count(body)}


def summarise(observations):
    """Counts over the sequence of rounds. Pure."""
    obs = list(observations or [])
    return {"rounds": len(obs),
            "flagged": sum(1 for o in obs if o.get("incomplete")),
            "item_counts": [o.get("items") for o in obs],
            "totals": [o.get("total") for o in obs]}


def counts_stable(observations):
    """Whether identical queries returned identical item counts. Pure."""
    counts = [o.get("items") for o in (observations or [])]
    return len(set(counts)) <= 1


def max_total(observations):
    """The largest reported match count across the rounds, or None. Pure."""
    totals = [o.get("total") for o in (observations or [])
              if o.get("total") is not None]
    return max(totals) if totals else None


def verdict(observations):
    """Classify the sequence. Pure. Returns (state, detail).

    Built from the sequence rather than from any one response, because a single
    flagged round cannot distinguish an unlucky query from a hopeless one, and
    those two have opposite repairs.
    """
    s = summarise(observations)
    if not s["rounds"]:
        return ("no-observations", "no round completed, so there is nothing to judge.")
    top = max_total(observations)
    ceiling = (" total_count is %s, well inside the %d-result ceiling, so the "
               "ceiling is not the explanation." % (top, RESULT_CAP)
               if top is not None and not above_result_cap(top) else "")
    if s["flagged"] and top is not None and above_result_cap(top):
        return ("timed-out-and-capped",
                "%d of %d round(s) came back partial and total_count is %s, "
                "which is also above the %d-result ceiling. These are two "
                "separate problems that look alike from outside and need "
                "repairing separately."
                % (s["flagged"], s["rounds"], top, RESULT_CAP))
    if s["flagged"] == s["rounds"]:
        return ("timed-out-always",
                "every one of %d round(s) came back partial, so this query does "
                "not finish inside the search timeout. No retry policy will fix "
                "that.%s" % (s["rounds"], ceiling))
    if s["flagged"]:
        return ("timed-out-intermittent",
                "%d of %d round(s) came back partial, so the query sometimes "
                "finishes and sometimes does not. A flagged response is a retry, "
                "not a result.%s" % (s["flagged"], s["rounds"], ceiling))
    if not counts_stable(observations):
        return ("unstable-counts",
                "no round was flagged, but identical queries returned %s "
                "item(s) across the rounds. Something is truncating or "
                "reordering underneath you, and the answer should be treated "
                "the same way as a flagged one."
                % " and ".join(str(c) for c in sorted(set(s["item_counts"]))))
    return ("complete",
            "%d of %d round(s) were unflagged and the item count did not move."
            % (s["rounds"], s["rounds"]))


def retry_or_narrow(observations):
    """What actually helps: retry, narrow, or nothing. Pure."""
    state = verdict(observations)[0]
    if state in ("timed-out-always", "timed-out-and-capped"):
        return "narrow"
    if state in ("timed-out-intermittent", "unstable-counts"):
        return "retry"
    return "nothing"


def repair(state, query=""):
    """The sentence a reader has to act on. Pure."""
    missing = ", ".join(narrowing(query)) or "nothing obvious"
    if state == "timed-out-always":
        return ("narrow the query until it finishes: add %s. Retrying will "
                "spend your search bucket on the same partial answer." % missing)
    if state == "timed-out-and-capped":
        return ("narrow the query until it finishes and until each slice reports "
                "under %d matches: add %s, then union the slices yourself."
                % (RESULT_CAP, missing))
    if state == "timed-out-intermittent":
        return ("treat a flagged response as a retry, never as a result, and "
                "never cache it. If the flag keeps coming back, add %s." % missing)
    if state == "unstable-counts":
        return ("treat this the same as a flagged response: do not cache it and "
                "do not diff against it. A moving count with no flag is still a "
                "moving count.")
    return "nothing."


def read_cost(queries, rounds):
    """Search requests this run will spend. Pure."""
    try:
        return max(0, len(queries or [])) * max(0, int(rounds))
    except (TypeError, ValueError):
        return 0


def within_search_bucket(cost):
    """Whether a plan of this size fits the per-minute search allowance. Pure."""
    try:
        return 0 < int(cost) <= SEARCH_BUCKET
    except (TypeError, ValueError):
        return False


def search(session, kind, query, per_page):
    """One search GET. Returns the decoded body or None."""
    r = session.get("%s/search/%s" % (API, kind),
                    params={"q": query, "per_page": per_page}, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from GitHub: GITHUB_TOKEN is missing, malformed "
                         "or revoked")
    if r.status_code == 403:
        raise SystemExit("403 from search. The search bucket is separate from "
                         "core and much tighter; wait for the window to reset")
    if r.status_code == 422:
        raise SystemExit("422 from search: the query is invalid, or you paged "
                         "past the %d-result ceiling" % RESULT_CAP)
    if r.status_code != 200:
        log.info("search returned %d", r.status_code)
        return None
    try:
        return r.json()
    except ValueError:
        return None


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--query", required=True, help="the search query, as sent")
    ap.add_argument("--kind", default="issues",
                    help="issues, repositories, code, users, commits")
    ap.add_argument("--rounds", type=int, default=3,
                    help="how many times to run the same query")
    ap.add_argument("--pause", type=float, default=2.0,
                    help="seconds between rounds")
    ap.add_argument("--per-page", type=int, default=100,
                    help="page size for each round")
    args = ap.parse_args()

    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        log.error("set GITHUB_TOKEN. Unauthenticated search gets a much smaller "
                  "bucket and this check would exhaust it")
        return 2

    cost = read_cost([args.query], args.rounds)
    if not within_search_bucket(cost):
        log.error("%d request(s) does not fit the %d per minute search bucket; "
                  "lower --rounds", cost, SEARCH_BUCKET)
        return 2
    log.info("read cost: %d search request(s) of the %d per minute search bucket",
             cost, SEARCH_BUCKET)

    session = requests.Session()
    session.headers.update({
        "Authorization": "Bearer " + token,
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        # GitHub rejects requests with no User-Agent outright.
        "User-Agent": UA,
    })

    observations = []
    for n in range(args.rounds):
        if n:
            time.sleep(args.pause)
        body = search(session, args.kind, args.query, args.per_page)
        if body is None:
            continue
        o = observe(body)
        observations.append(o)
        log.info("round %d: %d item(s), total_count %s, incomplete_results=%s",
                 n + 1, o["items"], o["total"], str(o["incomplete"]).lower())
        if not cacheable(body):
            log.info("round %d must not be cached or diffed against", n + 1)

    state, detail = verdict(observations)
    log.info("%s: %s", state, detail)
    missing = narrowing(args.query)
    if missing:
        log.info("missing from the query: %s", ", ".join(missing))
    log.info("what helps: %s", retry_or_narrow(observations))
    log.info("repair: %s", repair(state, args.query))

    print(json.dumps({"query": args.query,
                      "requests_spent": len(observations),
                      "summary": summarise(observations),
                      "counts_stable": counts_stable(observations),
                      "total_above_cap": above_result_cap(max_total(observations)),
                      "qualifiers_used": sorted(qualifiers(args.query)),
                      "narrowing_available": missing,
                      "action": retry_or_narrow(observations),
                      "state": state,
                      "detail": detail}, indent=2, default=str))
    return 1 if state not in ("complete", "no-observations") else 0


if __name__ == "__main__":
    sys.exit(main())

File: search-incomplete-results/node/github-search-incomplete.mjs
/**
 * Say whether a GitHub search is being answered in part and nobody noticed.
 *
 * Read only. One GET per round against /search/*, three rounds by default,
 * with a pause between them. Nothing is written.
 *
 * Search runs against a server-side timeout. When a query outruns it, GitHub
 * returns what it found with incomplete_results set to true rather than
 * failing. This is not the 1,000-result ceiling: total_count is read here
 * purely so that ceiling can be ruled out by name.
 *
 * Environment:
 *   GITHUB_TOKEN   any token with read access
 *   GITHUB_QUERY   the search query, as sent
 *   GITHUB_KIND    issues, repositories, code, users, commits. Default issues
 *   GITHUB_ROUNDS  how many times to run the same query. Default 3
 */
const API = 'https://api.github.com';
const UA = 'github-search-incomplete/1.0';

/** The other Search limit, read only so it can be excluded as an explanation. */
export const RESULT_CAP = 1000;
/** Authenticated search requests per minute. The check is sized against this. */
export const SEARCH_BUCKET = 30;

const QUALIFIER = /(?:^|\s)-?([A-Za-z_]+):\S/g;

export const SCOPES = ['repo', 'org', 'user'];
export const RANGES = ['created', 'updated', 'merged', 'closed'];

/** Whether this response says it is partial. Pure. */
export function flagged(body) {
  return Boolean(body && typeof body === 'object' && body.incomplete_results === true);
}

/** The reported match count, or null. Pure. */
export function totalOf(body) {
  if (!body || typeof body !== 'object') return null;
  const raw = body.total_count;
  if (raw === null || raw === undefined || raw === '') return null;
  const n = Number(raw);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

/** How many items were actually delivered. Pure. */
export function itemCount(body) {
  if (!body || typeof body !== 'object') return 0;
  return Array.isArray(body.items) ? body.items.length : 0;
}

/** Whether this response may be stored. Pure. */
export function cacheable(body) {
  return Boolean(body && typeof body === 'object') && !flagged(body);
}

/** Whether the ceiling could also be in play here. Pure. */
export function aboveResultCap(total) {
  const n = Number(total);
  return Number.isFinite(n) && n > RESULT_CAP;
}

/** The qualifier names used in a search query. Pure. */
export function qualifiers(query) {
  const out = new Set();
  for (const m of ` ${String(query ?? '')}`.matchAll(QUALIFIER)) out.add(m[1]);
  return out;
}

/** Which narrowing devices the query is not already using. Pure. */
export function narrowing(query) {
  const have = qualifiers(query);
  const out = [];
  if (!SCOPES.some((s) => have.has(s))) out.push('repo: or org:');
  if (!RANGES.some((s) => have.has(s))) out.push('created: or updated: date range');
  if (!have.has('language')) out.push('language:');
  return out;
}

/** The three fields worth keeping from one response. Pure. */
export function observe(body) {
  return { incomplete: flagged(body), total: totalOf(body), items: itemCount(body) };
}

/** Counts over the sequence of rounds. Pure. */
export function summarise(observations) {
  const obs = Array.isArray(observations) ? observations : [];
  return {
    rounds: obs.length,
    flagged: obs.filter((o) => o && o.incomplete).length,
    item_counts: obs.map((o) => (o || {}).items),
    totals: obs.map((o) => (o || {}).total),
  };
}

/** Whether identical queries returned identical item counts. Pure. */
export function countsStable(observations) {
  const counts = (Array.isArray(observations) ? observations : []).map((o) => (o || {}).items);
  return new Set(counts).size <= 1;
}

/** The largest reported match count across the rounds, or null. Pure. */
export function maxTotal(observations) {
  const totals = (Array.isArray(observations) ? observations : [])
    .map((o) => (o || {}).total).filter((t) => t !== null && t !== undefined);
  return totals.length ? Math.max(...totals) : null;
}

/** Classify the sequence. Pure. Returns [state, detail]. */
export function verdict(observations) {
  const s = summarise(observations);
  if (!s.rounds) return ['no-observations', 'no round completed, so there is nothing to judge.'];
  const top = maxTotal(observations);
  const ceiling = top !== null && !aboveResultCap(top)
    ? ` total_count is ${top}, well inside the ${RESULT_CAP}-result ceiling, so the ceiling is not the explanation.`
    : '';
  if (s.flagged && top !== null && aboveResultCap(top)) {
    return ['timed-out-and-capped',
      `${s.flagged} of ${s.rounds} round(s) came back partial and total_count is `
      + `${top}, which is also above the ${RESULT_CAP}-result ceiling. These are `
      + 'two separate problems that look alike from outside and need repairing separately.'];
  }
  if (s.flagged === s.rounds) {
    return ['timed-out-always',
      `every one of ${s.rounds} round(s) came back partial, so this query does not `
      + `finish inside the search timeout. No retry policy will fix that.${ceiling}`];
  }
  if (s.flagged) {
    return ['timed-out-intermittent',
      `${s.flagged} of ${s.rounds} round(s) came back partial, so the query `
      + 'sometimes finishes and sometimes does not. A flagged response is a '
      + `retry, not a result.${ceiling}`];
  }
  if (!countsStable(observations)) {
    const seen = [...new Set(s.item_counts)].sort((a, b) => a - b).join(' and ');
    return ['unstable-counts',
      `no round was flagged, but identical queries returned ${seen} item(s) across `
      + 'the rounds. Something is truncating or reordering underneath you, and the '
      + 'answer should be treated the same way as a flagged one.'];
  }
  return ['complete',
    `${s.rounds} of ${s.rounds} round(s) were unflagged and the item count did not move.`];
}

/** What actually helps: retry, narrow, or nothing. Pure. */
export function retryOrNarrow(observations) {
  const state = verdict(observations)[0];
  if (['timed-out-always', 'timed-out-and-capped'].includes(state)) return 'narrow';
  if (['timed-out-intermittent', 'unstable-counts'].includes(state)) return 'retry';
  return 'nothing';
}

/** The sentence a reader has to act on. Pure. */
export function repair(state, query = '') {
  const missing = narrowing(query).join(', ') || 'nothing obvious';
  if (state === 'timed-out-always') {
    return `narrow the query until it finishes: add ${missing}. Retrying will `
      + 'spend your search bucket on the same partial answer.';
  }
  if (state === 'timed-out-and-capped') {
    return 'narrow the query until it finishes and until each slice reports under '
      + `${RESULT_CAP} matches: add ${missing}, then union the slices yourself.`;
  }
  if (state === 'timed-out-intermittent') {
    return 'treat a flagged response as a retry, never as a result, and never '
      + `cache it. If the flag keeps coming back, add ${missing}.`;
  }
  if (state === 'unstable-counts') {
    return 'treat this the same as a flagged response: do not cache it and do not '
      + 'diff against it. A moving count with no flag is still a moving count.';
  }
  return 'nothing.';
}

/** Search requests this run will spend. Pure. */
export function readCost(queries, rounds) {
  const q = Array.isArray(queries) ? queries.length : 0;
  const r = Number(rounds);
  return q * (Number.isFinite(r) && r > 0 ? Math.trunc(r) : 0);
}

/** Whether a plan of this size fits the per-minute search allowance. Pure. */
export function withinSearchBucket(cost) {
  const n = Number(cost);
  return Number.isFinite(n) && n > 0 && n <= SEARCH_BUCKET;
}

function headers(token) {
  return {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
    'User-Agent': UA,
  };
}

const wait = (ms) => new Promise((resolve) => { setTimeout(resolve, ms); });

async function main() {
  const token = (process.env.GITHUB_TOKEN || "dummy-github-token");
  const query = (process.env.GITHUB_QUERY || "dummy-github-query");
  if (!token || !query) {
    console.error('set GITHUB_TOKEN (read-only is enough) and GITHUB_QUERY');
    process.exitCode = 2;
    return;
  }
  const kind = (process.env.GITHUB_KIN || "dummy-github-kin")D || 'issues';
  const rounds = Number((process.env.GITHUB_ROUND || "dummy-github-round")S || 3);
  const cost = readCost([query], rounds);
  if (!withinSearchBucket(cost)) {
    console.error(`${cost} request(s) does not fit the ${SEARCH_BUCKET} per minute search bucket`);
    process.exitCode = 2;
    return;
  }
  console.log(`read cost: ${cost} search request(s) of the ${SEARCH_BUCKET} per minute search bucket`);

  const observations = [];
  for (let n = 0; n < rounds; n += 1) {
    if (n) await wait(2000);
    const url = new URL(`${API}/search/${kind}`);
    url.searchParams.set('q', query);
    url.searchParams.set('per_page', '100');
    const res = await fetch(url, { headers: headers(token) });
    if (res.status !== 200) {
      console.error(`search returned ${res.status}`);
      continue;
    }
    let body = null;
    try { body = await res.json(); } catch { body = null; }
    if (body === null) continue;
    const o = observe(body);
    observations.push(o);
    console.log(`round ${n + 1}: ${o.items} item(s), total_count ${o.total}, `
      + `incomplete_results=${o.incomplete}`);
    if (!cacheable(body)) console.log(`round ${n + 1} must not be cached or diffed against`);
  }

  const [state, detail] = verdict(observations);
  console.log(`${state}: ${detail}`);
  const missing = narrowing(query);
  if (missing.length) console.log(`missing from the query: ${missing.join(', ')}`);
  console.log(`what helps: ${retryOrNarrow(observations)}`);
  console.log(`repair: ${repair(state, query)}`);

  console.log(JSON.stringify({
    query,
    requests_spent: observations.length,
    summary: summarise(observations),
    counts_stable: countsStable(observations),
    total_above_cap: aboveResultCap(maxTotal(observations)),
    qualifiers_used: [...qualifiers(query)].sort(),
    narrowing_available: missing,
    action: retryOrNarrow(observations),
    state,
    detail,
  }, null, 2));
  process.exitCode = ['complete', 'no-observations'].includes(state) ? 0 : 1;
}

// Guarded so importing this file from the test runner does not start a run and
// fail the suite even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
